### DISCUSSION TITLE
Watch Component setValue Controlled Component

### DISCUSSION THREAD

**Discussion by bjrice88** [react-hook-form/react-hook-form#7744] [Q&A]
I'm running into a React error in a very specific situation and I'm not sure if it is actually an issue.  I have a component in my form context that watches a few form values and in certain situations I want a form value to be disabled and the value to be controlled programmatically.

Functionally this works perfectly as I would expect.  I have a react component that watches some values, if the values match a criteria then it sets the value of another form value programmatically.

The only problem is that in the console I get a warning the first time that the setValue is triggered `Warning: Cannot update a component (`Unknown`) while rendering a different component (`WatchCheckbox`). To locate the bad setState() call inside `WatchCheckbox`, follow the stack trace as described`

The strange thing is that it only happens on the first time that setValue is called, and even then it works as expected.  Am I doing something terribly wrong here?

I've attached a codesandbox with a demonstration.  Basically in the demo I want the textbox to reflect the value of the select when the checkbox is true, but not when it's false.

https://codesandbox.io/s/react-hook-form-v7-controller-forked-8eujk

**Comment by Moshyfawn** [comment]
The React warning is because you need to call your state setters _(`setValue` in case of RHF)_ inside the `useEffect` on re-render. [CSB without the warning](https://codesandbox.io/s/react-hook-form-v7-controller-forked-qgo0c?file=/src/index.js)

**Reply by chouclee** [reply]
Hit the same issue. `useEffect` can hide the warning, but doesn't solve the rendering issue. 

- useEffect shouldn't be used on watched field: https://github.com/react-hook-form/react-hook-form/issues/7068#issuecomment-972162035

- The intention if `setValue` is to update the value **before** current rendering so UI could render the correct value. It may cause other issues if `setValue` is triggered within `useEffect`, e.g. for my use case, I need to clear the selected value when dropdown items are changed. Putting `setValue` within `useEffect` may receive warnings from some UI libraries: `You have provided an out-of-range value foo for the select component`
